crypto/random/rand.go

```diff
@@ -149,11 +149,7 @@ func (p *genericPRG) Shuffle(n int, swap func(i, j int)) error {
 	if n < 0 {
 		return fmt.Errorf("population size cannot be negative")
 	}
-	for i := n - 1; i > 0; i-- {
```

Comment:
the algorithms in utils/rand/rand.go is basically identical to this.  the only difference is the random source.  can you refactor the code to share implementation?

btw, Permutation / SubPermutation should also just call Shuffle/Sample

maybe undo changes to this file in this PR (since it's unrelated to test updates), and make the changes in a separate PR

Reply:
> Permutation / SubPermutation should also just call Shuffle/Sample

This is true, though it is slightly less efficient: in order to use `Shuffle`, `Permutation` needs to initialize an array firts, and then shuffle it. The current implementation of `Permutation` initializes and permutes at the same time (it's an optimized [variant](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_%22inside-out%22_algorithm) of Fisher-Yates).

> can you refactor the code to share implementation?

You are right that the core idea is the same. I did explore the idea of combining both implementations but I didn't find an elegant way to do that, given there are minor differences between the packages (returned errors, thread safety, integer types). For now I only combined the test tools, but we can discuss ways to combine the rest.